lua/nvim-treesitter/utils.lua

```diff
@@ -37,6 +46,49 @@ function M.get_cache_dir()
   return nil, 'Invalid cache rights, '..fn.stdpath('data')..' or /tmp should be read/write'
 end
 
+-- Returns $XDG_DATA_HOME/nvim/site, but could use any directory that is in
+-- runtimepath
+function M.get_site_dir()
+  local path_sep = M.get_path_sep()
+  return fn.stdpath('data')..path_sep..'site'
+end
+
+-- Try the package dir of the nvim-treesitter plugin first, followed by the
+-- "site" dir from "runtimepath". "site" dir will be created if it doesn't
+-- exist. Using only the package dir won't work when the plugin is installed
+-- with Nix, since the "/nix/store" is read-only.
+function M.get_parser_install_dir()
+  local package_path = M.get_package_path()
+
+  -- If package_path is read/write, use that
+  if luv.fs_access(package_path, 'RW') then
+    return package_path
+  end
+
+  local site_dir = M.get_site_dir()
+  local path_sep = M.get_path_sep()
+  local parser_dir = site_dir..path_sep..'parser'
+
+  -- Try creating and using parser_dir if it doesn't exist
+  if not luv.fs_stat(parser_dir) then
+    local ok, error = pcall(vim.fn.mkdir, parser_dir, "p", "0755")
+    if not ok then
+      return nil, 'Couldn\'t create parser dir '..parser_dir..': '..error
+    end
+
+    return parser_dir
+  end
+
+  -- parser_dir exists, use it if it's read/write
+  if luv.fs_access(parser_dir, 'RW') then
+    return parser_dir
+  end
+
+  -- package_path isn't read/write, parser_dir exists but isn't read/write
+  -- either, give up
+  return nil, 'Invalid cache rights, '..package_path..' or '..parser_dir..' should be read/write'
+end
```

Comment:
yep i agree with @vigoux, we should have a function like
```lua
local function genJoin()
  local path_sep = get_path_sep()
  return function(...)
    return ....
  end
end
function M.join(...) = genJoin()
```

Reply:
Will address the feedback tomorrow, thanks for the helpful suggestions! 🎊

Reply:
*gen_join (kyazdani were not doing JS here ;-))